alexandria/encoding/src/base64.cairo

```diff
@@ -0,0 +1,326 @@
+use core::clone::Clone;
+use array::ArrayTrait;
+use integer::{upcast, downcast};
+use option::OptionTrait;
+use debug::PrintTrait;
+use alexandria_encoding::math::{shl, shr, U6_MAX, U8_MAX};
+
+trait Encoder<T> {
+    fn encode(data: Array<T>) -> Array<u8>;
+}
+
+trait Decoder<T> {
+    fn decode(data: Array<T>) -> Array<u8>;
+}
+
+impl u8Base64Encoder of Encoder<u8> {
+    fn encode(mut data: Array<u8>) -> Array<u8> {
+        let mut p = if (data.len() % 3 == 1) {
+            data.append(0);
+            data.append(0);
+            2
+        } else if (data.len() % 3 == 2) {
+            data.append(0);
+            1
+        } else {
+            0
+        };
+
+        let mut char_set = get_base64_char_set();
+        let mut result = ArrayTrait::new();
+        encode_loop(p, ref data, 0, ref char_set, ref result);
+        result
+    }
+}
+
+impl u8Base64UrlEncoder of Encoder<u8> {
+    fn encode(mut data: Array<u8>) -> Array<u8> {
+        let mut p = if (data.len() % 3 == 1) {
+            data.append(0);
+            data.append(0);
+            2
+        } else if (data.len() % 3 == 2) {
+            data.append(0);
+            1
+        } else {
+            0
+        };
+
+        let mut char_set = get_base64_url_char_set();
+        let mut result = ArrayTrait::new();
+        encode_loop(p, ref data, 0, ref char_set, ref result);
+        result
+    }
+}
+
+impl u8Base64Decoder of Decoder<u8> {
+    fn decode(mut data: Array<u8>) -> Array<u8> {
+        let mut result = ArrayTrait::new();
+        let mut p = 0_u8;
+        if data.len() > 0 {
+            if *data[data.len() - 1] == '=' {
+                p += 1;
+            }
+            if *data[data.len() - 2] == '=' {
+                p += 1;
+            }
+            decode_loop(p, ref data, 0, ref result);
+        }
+        result
+    }
+}
+
+impl u8Base64UrlDecoder of Decoder<u8> {
```

Comment:
As both decoder are the same, I wouldn't do two just for the sake of having one that suit each name.
But if you think it makes sense I guess we can keep it, no strong opinion